src/backend/libinput.rs

```diff
@@ -0,0 +1,147 @@
+
+                            let new_caps = SeatCapabilities {
+                                pointer: added.has_capability(DeviceCapability::Pointer),
+                                keyboard: added.has_capability(DeviceCapability::Keyboard),
+                                touch: added.has_capability(DeviceCapability::Touch),
+                            };
+
+                            let device_seat = added.seat();
+                            self.devices.push(added);
+
+                            let contains = self.seats.contains_key(&device_seat);
+                            if contains {
+                                let old_seat = self.seats.get_mut(&device_seat).unwrap();
+                                {
+                                    let caps = old_seat.capabilities_mut();
+                                    caps.pointer = new_caps.pointer || caps.pointer;
+                                    caps.keyboard = new_caps.keyboard || caps.keyboard;
+                                    caps.touch = new_caps.touch || caps.touch;
+                                }
+                                if let Some(ref mut handler) = self.handler {
+                                    handler.on_seat_changed(old_seat);
+                                }
+                            } else {
+                                let mut hasher = DefaultHasher::default();
+                                device_seat.hash(&mut hasher);
+                                self.seats.insert(device_seat.clone(), Seat::new(hasher.finish(), new_caps));
+                                if let Some(ref mut handler) = self.handler {
+                                    handler.on_seat_created(self.seats.get(&device_seat).unwrap());
+                                }
+                            }
+                        },
+                        DeviceEvent::Removed(device_removed_event) => {
+                            let removed = device_removed_event.into_event().device();
+
+                            // remove device
+                            self.devices.retain(|dev| *dev == removed);
+
+                            let device_seat = removed.seat();
+
+                            // update capabilities, so they appear correctly on `on_seat_changed` and `on_seat_destroyed`.
+                            if let Some(seat) = self.seats.get_mut(&device_seat) {
+                                let caps = seat.capabilities_mut();
+                                caps.pointer = self.devices.iter().any(|x| x.has_capability(DeviceCapability::Pointer));
+                                caps.keyboard = self.devices.iter().any(|x| x.has_capability(DeviceCapability::Keyboard));
+                                caps.touch = self.devices.iter().any(|x| x.has_capability(DeviceCapability::Touch));
+                            } else {
+                                //TODO is this the best we can do?
```

Comment:
Seems like the tight think to do in my opinion. This case would be a bug (be it in smithay, input.rs or libinput), and we want to catch it clearly rather than silently try to handle it.